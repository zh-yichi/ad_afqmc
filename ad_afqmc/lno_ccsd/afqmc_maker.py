from jax import numpy as jnp
from jax import vmap

def walker_norhf_energy(walker,ham_data,trial_mo):
    '''rhf walker's energy with non-orthonal trial'''
    rot_h1,rot_chol = ham_data['rot_h1'],ham_data['rot_chol']
    nocc = rot_h1.shape[0]
    green_walker = (walker.dot(jnp.linalg.inv(trial_mo.T.conj() @ walker))).T
    f = jnp.einsum('gia,ak->gik',rot_chol[:,:nocc,nocc:],
                   green_walker.T[nocc:,:nocc],optimize='optimal')
    tf = vmap(jnp.trace)(f)
    tc = vmap(jnp.trace)(rot_chol[:,:nocc,:nocc])
    ene11 = 2*jnp.einsum('ia,ia->',green_walker[:nocc,nocc:],rot_h1[:nocc,nocc:])
    ene12 = 2*jnp.sum(tf*tc) \
            - jnp.einsum('gik,gki->',f,rot_chol[:,:nocc,:nocc])
    ene1 = ene11-2*ene12
    ene2 = 2*jnp.sum(tf*tf) \
           - jnp.einsum('gij,gji->',f,f)
    return ene1+ene2 #,ene1,ene2,

def walker_norhf_orb_en(walker,ham_data,wave_data):
    '''rhf walker's energy with non-orthonal trial'''
    rot_h1,rot_chol = ham_data['rot_h1'],ham_data['rot_chol']
    m = jnp.dot(wave_data["prjlo"].T,wave_data["prjlo"])
    nocc = rot_h1.shape[0]
    green_walker = (walker.dot(jnp.linalg.inv(
        wave_data['mo_coeff'].T.conj() @ walker))).T
    f = jnp.einsum('gia,ak->gik',rot_chol[:,:nocc,nocc:],
                   green_walker.T[nocc:,:nocc],optimize='optimal')
    tf = vmap(jnp.trace)(f)
    # tc = vmap(jnp.trace)(rot_chol[:,:nocc,:nocc])
    ene11 = 2*jnp.einsum('ia,ka,ik->',green_walker[:nocc,nocc:],rot_h1[:nocc,nocc:],m)
    ene12 = 2*jnp.einsum('gik,ik,gjj->',f,m,rot_chol[:,:nocc,:nocc]) \
            - jnp.einsum('gij,gjk,ik->',f,rot_chol[:,:nocc,:nocc],m)
    ene1 = ene11-2*ene12
    ene2 = 2*jnp.einsum('gik,ik,g->',f,m,tf) \
           - jnp.einsum('gij,gjk,ik->',f,f,m)
    return ene1+ene2