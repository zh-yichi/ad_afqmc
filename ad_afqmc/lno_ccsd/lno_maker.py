import numpy as np

def thouless_trans(t1):
    q, r = np.linalg.qr(t1)
    u_ai = r.T
    u_ji = q
    u_occ = np.vstack((u_ji,u_ai))
    u, _, _ = np.linalg.svd(u_occ)
    return u
