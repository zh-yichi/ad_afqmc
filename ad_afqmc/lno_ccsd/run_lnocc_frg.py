from functools import partial
from jax import random
#from mpi4py import MPI
import numpy as np
from jax import numpy as jnp
from ad_afqmc import config, wavefunctions, stat_utils
from ad_afqmc.lno_ccsd import lno_ccsd
import time

print = partial(print, flush=True)

ham_data, ham, prop, trial, wave_data, sampler, observable, options, _ = (
    lno_ccsd.prep_lnoccsd_afqmc())

if options["use_gpu"]:
    config.afqmc_config["use_gpu"] = True

config.setup_jax()
MPI = config.setup_comm()

# init_time = time.time()
comm = MPI.COMM_WORLD
rank = comm.Get_rank()  # Process rank
size = comm.Get_size()  # Total number of processes


### initialize propagation
seed = options["seed"]
propagator = prop
init_walkers = None
ham_data = wavefunctions.rhf(trial.norb, trial.nelec,n_batch=trial.n_batch
                                )._build_measurement_intermediates(ham_data, wave_data)
ham_data = ham.build_measurement_intermediates(ham_data, trial, wave_data)
ham_data = ham.build_propagation_intermediates(ham_data, propagator, trial, wave_data)

prop_data = propagator.init_prop_data(trial, wave_data, ham_data, init_walkers)
if jnp.abs(jnp.sum(prop_data["overlaps"])) < 1.0e-6:
    raise ValueError(
        "Initial overlaps are zero. Pass walkers with non-zero overlap."
    )
prop_data["key"] = random.PRNGKey(seed + rank)

prop_data["overlaps"] = trial.calc_overlap(prop_data["walkers"], wave_data)
prop_data["n_killed_walkers"] = 0
prop_data["pop_control_ene_shift"] = prop_data["e_estimate"]

### relaxation ###
init_time = time.time()
comm.Barrier()
if rank == 0:
    hf_orb_cr,hf_orb_en = lno_ccsd._frg_hf_cr(
        ham_data['rot_h1'], ham_data['rot_chol'],prop_data["walkers"][0],trial,wave_data)
    cc_orb_cr = lno_ccsd._frg_ccsd_cr(
        prop_data["walkers"][0],ham_data,wave_data,trial,1e-6)
    
    cc_orb_en = hf_orb_cr+cc_orb_cr
    e_init = prop_data["e_estimate"]

    print(f'# afqmc propagation with {options["n_walkers"]*size} walkers')
    print(f'# appraching equilibrium')
    print('# step  energy  hf_orb_en  cc_orb_en  time')
    print(f"  {0:3d}"
          f"  {e_init:.6f}"
          f"  {hf_orb_en:.6f}"
          f"  {cc_orb_en:.6f}"
          f"  {time.time() - init_time:.2f} "
        )
comm.Barrier()

for n in range(options["n_eql"]):
    prop_data, (blk_en,blk_wt,blk_hf_orb_en,blk_cc_orb_en) \
                = lno_ccsd.propagate_phaseless_orb(
                    ham_data,prop,prop_data,trial,wave_data,sampler)
    
    blk_en = np.array([blk_en], dtype="float32")
    blk_wt = np.array([blk_wt], dtype="float32")
    blk_hf_orb_en = np.array([blk_hf_orb_en], dtype="float32")
    blk_cc_orb_en = np.array([blk_cc_orb_en], dtype="float32")    

    blk_wt_en = np.array(
        [blk_en * blk_wt], dtype="float32"
    )
    blk_wt_hf_orb_en = np.array(
        [blk_hf_orb_en * blk_wt], dtype="float32"
    )
    blk_wt_cc_orb_en = np.array(
        [blk_cc_orb_en * blk_wt], dtype="float32"
    )

    tot_wt_en = np.zeros(1, dtype="float32")
    tot_wt = np.zeros(1, dtype="float32")
    tot_wt_hf_orb_en = np.zeros(1, dtype="float32")
    tot_wt_cc_orb_en = np.zeros(1, dtype="float32")

    comm.Reduce(
            [blk_wt_en, MPI.FLOAT],
            [tot_wt_en, MPI.FLOAT],
            op=MPI.SUM,
            root=0,
        )
    comm.Reduce(
        [blk_wt, MPI.FLOAT],
        [tot_wt, MPI.FLOAT],
        op=MPI.SUM,
        root=0,
    )
    comm.Reduce(
            [blk_wt_hf_orb_en, MPI.FLOAT],
            [tot_wt_hf_orb_en, MPI.FLOAT],
            op=MPI.SUM,
            root=0,
        )
    comm.Reduce(
            [blk_wt_cc_orb_en, MPI.FLOAT],
            [tot_wt_cc_orb_en, MPI.FLOAT],
            op=MPI.SUM,
            root=0,
        )

    comm.Barrier()
    if rank == 0:
        blk_en = tot_wt_en / tot_wt
        blk_wt = tot_wt
        blk_hf_orb_en = tot_wt_hf_orb_en / tot_wt
        blk_cc_orb_en = tot_wt_cc_orb_en / tot_wt

    comm.Bcast(blk_en, root=0)
    comm.Bcast(blk_wt, root=0)
    comm.Bcast(blk_hf_orb_en, root=0)
    comm.Bcast(blk_cc_orb_en, root=0)
    
    prop_data = propagator.orthonormalize_walkers(prop_data)
    prop_data = propagator.stochastic_reconfiguration_global(prop_data, comm)
    prop_data["e_estimate"] = (
         0.9 * prop_data["e_estimate"] + 0.1 * blk_en[0]
         )
    comm.Barrier()

    comm.Barrier()
    if rank == 0:
        print(
            f"  {n+1:3d}"
            f"  {blk_en[0]:.6f}"
            f"  {blk_hf_orb_en[0]:.6f}"
            f"  {blk_cc_orb_en[0]:.6f}"
            f"  {time.time() - init_time:.2f} "
        )
    comm.Barrier()

glb_blk_en = None
glb_blk_wt = None
glb_blk_hf_orb_en = None
glb_blk_cc_orb_en = None

if rank == 0:
    glb_blk_en = np.zeros(size * sampler.n_blocks)
    glb_blk_wt = np.zeros(size * sampler.n_blocks)
    glb_blk_hf_orb_en = np.zeros(size * sampler.n_blocks)
    glb_blk_cc_orb_en = np.zeros(size * sampler.n_blocks)
    
comm.Barrier()
if rank == 0:
    print("# Sampling sweeps:")
    print("# iter  energy  err  hf_orb_en  err  cc_orb_cr  err  time")
comm.Barrier()

for n in range(sampler.n_blocks):
    prop_data,(blk_en,blk_wt,blk_hf_orb_en,blk_cc_orb_en) \
        = lno_ccsd.propagate_phaseless_orb(
            ham_data,prop,prop_data,trial,wave_data,sampler)

    blk_en = np.array([blk_en], dtype="float32")
    blk_wt = np.array([blk_wt], dtype="float32")
    blk_hf_orb_en = np.array([blk_hf_orb_en], dtype="float32")
    blk_cc_orb_en = np.array([blk_cc_orb_en], dtype="float32")

    gather_en = None
    gather_wt = None
    gather_hf_orb_en = None
    gather_cc_orb_en = None

    comm.Barrier()
    if rank == 0:
        gather_en = np.zeros(size, dtype="float32")
        gather_wt = np.zeros(size, dtype="float32")
        gather_hf_orb_en = np.zeros(size, dtype="float32")
        gather_cc_orb_en = np.zeros(size, dtype="float32")
    comm.Barrier()

    comm.Gather(blk_en, gather_en, root=0)
    comm.Gather(blk_wt, gather_wt, root=0)
    comm.Gather(blk_hf_orb_en, gather_hf_orb_en, root=0)
    comm.Gather(blk_cc_orb_en, gather_cc_orb_en, root=0)

    comm.Barrier()
    if rank == 0:
        glb_blk_en[n * size : (n + 1) * size] = gather_en
        glb_blk_wt[n * size : (n + 1) * size] = gather_wt
        glb_blk_hf_orb_en[n * size : (n + 1) * size] = gather_hf_orb_en
        glb_blk_cc_orb_en[n * size : (n + 1) * size] = gather_cc_orb_en

        assert gather_wt is not None

        blk_wt= np.sum(gather_wt)
        blk_en = np.sum(gather_wt * gather_en) / blk_wt
    comm.Barrier()
    
    prop_data = propagator.orthonormalize_walkers(prop_data)
    prop_data = propagator.stochastic_reconfiguration_global(prop_data, comm)
    prop_data["e_estimate"] = 0.9 * prop_data["e_estimate"] + 0.1 * blk_en

    if n % (max(sampler.n_blocks // 10, 1)) == 0:
        comm.Barrier()
        if rank == 0:
                en, en_err = stat_utils.blocking_analysis(
                    glb_blk_wt[: (n + 1) * size],
                    glb_blk_en[: (n + 1) * size],
                    neql=0,
                )
                hf_orb_en, hf_orb_en_err = stat_utils.blocking_analysis(
                    glb_blk_wt[: (n + 1) * size],
                    glb_blk_hf_orb_en[: (n + 1) * size],
                    neql=0,
                )
                cc_orb_en, cc_orb_en_err = stat_utils.blocking_analysis(
                    glb_blk_wt[: (n + 1) * size],
                    glb_blk_cc_orb_en[: (n + 1) * size],
                    neql=0,
                )

                if en_err is not None:
                    en_err = f"{en_err:.6f}"
                else:
                    en_err = f"  {en_err}  "
                if hf_orb_en_err is not None:
                    hf_orb_en_err = f"{hf_orb_en_err:.6f}"
                else:
                    hf_orb_en_err = f"  {hf_orb_en_err}  "
                if cc_orb_en_err is not None:
                    cc_orb_en_err = f"{cc_orb_en_err:.6f}"
                else:
                    cc_orb_en_err = f"  {cc_orb_en_err}  "

                en = f"{en:.6f}"
                hf_orb_en = f"{hf_orb_en:.6f}"
                cc_orb_en = f"{cc_orb_en:.6f}"

                print(f"  {n:4d}  {en}  {en_err}  {hf_orb_en}  {hf_orb_en_err}"
                      f"  {cc_orb_en}  {cc_orb_en_err}  {time.time() - init_time:.2f}")
        comm.Barrier()


comm.Barrier()
if rank == 0:
    assert glb_blk_en is not None
    assert glb_blk_wt is not None
    assert glb_blk_hf_orb_en is not None
    assert glb_blk_cc_orb_en is not None


    samples_clean, idx = stat_utils.reject_outliers(
    np.stack(
                (
                    glb_blk_wt,
                    glb_blk_en,
                    glb_blk_hf_orb_en,
                    glb_blk_cc_orb_en,
                )
            ).T,
            1,
        )

    print(
        f"# Number of outliers in post: {glb_blk_wt.size - samples_clean.shape[0]} "
        )
    
    glb_blk_wt = samples_clean[:, 0]
    glb_blk_en = samples_clean[:, 1]
    glb_blk_hf_orb_en = samples_clean[:, 2]
    glb_blk_cc_orb_en = samples_clean[:, 3]

    en, en_err = stat_utils.blocking_analysis(
                    glb_blk_wt[: (n + 1) * size],
                    glb_blk_en[: (n + 1) * size],
                    neql=0,printQ=True
                )
    hf_orb_en, hf_orb_en_err = stat_utils.blocking_analysis(
                    glb_blk_wt[: (n + 1) * size],
                    glb_blk_hf_orb_en[: (n + 1) * size],
                    neql=0,printQ=True
                )
    cc_orb_en, cc_orb_en_err = stat_utils.blocking_analysis(
                    glb_blk_wt[: (n + 1) * size],
                    glb_blk_cc_orb_en[: (n + 1) * size],
                    neql=0,printQ=True
                )
    
    if en_err is not None:
        en_err = f"{en_err:.6f}"
    else:
        en_err = f"  {en_err}  "
    if hf_orb_en_err is not None:
        hf_orb_en_err = f"{hf_orb_en_err:.6f}"
    else:
        hf_orb_en_err = f"  {hf_orb_en_err}  "
    if cc_orb_en_err is not None:
        cc_orb_en_err = f"{cc_orb_en_err:.6f}"
    else:
        cc_orb_en_err = f"  {cc_orb_en_err}  "


    en = f"{en:.6f}"
    hf_orb_en = f"{hf_orb_en:.6f}"
    cc_orb_en = f"{cc_orb_en:.6f}"

    print(f"Final Results:")
    print(f"lno-ccsd-afqmc energy: {en} +/- {en_err}")
    print(f"lno-ccsd-afqmc hf_orb_en: {hf_orb_en} +/- {hf_orb_en_err}")
    print(f"lno-ccsd-afqmc cc_orb_en: {cc_orb_en} +/- {cc_orb_en_err}")
    print(f"total run time: {time.time() - init_time:.2f}")

comm.Barrier()
