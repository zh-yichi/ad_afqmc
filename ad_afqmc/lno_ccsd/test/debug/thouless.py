from pyscf import gto, scf, cc
from ad_afqmc.lno_ccsd import lno_maker

a = 1. # 2aB
nH = 6
atoms = ""
for i in range(nH):
    atoms += f"H {i*a:.5f} 0.00000 0.00000 \n"

mol = gto.M(atom=atoms, basis="sto6g", verbose=4)
mf = scf.RHF(mol)
mf.kernel()

mycc = cc.CCSD(mf)
mycc.kernel()

# from mpi4py import MPI
# MPI.Finalize()

options = {'n_eql': 5,
           'n_prop_steps': 30,
            'n_ene_blocks': 1,
            'n_sr_blocks': 10,
            'n_blocks': 5,
            'n_walkers': 1,
            'seed': 98,
            'walker_type': 'rhf',
            'trial': 'rhf',
            'dt':0.005,
            'ad_mode':None,
            'use_gpu': False,
            }

u = lno_maker.thouless_trans(mycc.t1)
mo_t = mf.mo_coeff @ u
mf.mo_coeff = mo_t

from ad_afqmc import pyscf_interface, run_afqmc
from mpi4py import MPI
if not MPI.Is_finalized():
    MPI.Finalize()

pyscf_interface.prep_afqmc(mf,chol_cut=1e-7)
run_afqmc.run_afqmc(options=options, nproc=5)
# ham_data, ham, prop, trial, wave_data, sampler, observable, options, _ \
#     = (mpi_jax._prep_afqmc(options))
# nwalkers = 1
# init_walkers = np.stack([wave_data["mo_coeff"]] * nwalkers, axis=0)
# wave_data["mo_coeff"] = wave_data["mo_coeff"] @ u
# driver.afqmc(ham_data, ham, prop, trial, wave_data, sampler, observable, options, MPI,init_walkers)