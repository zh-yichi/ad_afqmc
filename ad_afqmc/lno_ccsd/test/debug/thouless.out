#INFO: **** input file is /home/yichi/research/software/cs_afqmc/ad_afqmc/lno_ccsd/test/debug/thouless.py ****
from pyscf import gto, scf, cc
from ad_afqmc.lno_ccsd import lno_maker
import numpy as np

a = 1. # 2aB
nH = 10
atoms = ""
for i in range(nH):
    atoms += f"H {i*a:.5f} 0.00000 0.00000 \n"

mol = gto.M(atom=atoms, basis="sto6g", verbose=4)
mf = scf.RHF(mol)
mf.kernel()

mycc = cc.CCSD(mf)
mycc.kernel()

t2_0 = np.zeros(mycc.t2.shape)
eris = mycc.ao2mo(mycc.mo_coeff)
eccsd = mycc.energy(mycc.t1, t2_0, eris)
ecc_t2_0 = eccsd+mf.e_tot
print(f'ccsd energy with T2=0 is:{ecc_t2_0}')

# from mpi4py import MPI
# MPI.Finalize()

options = {'n_eql': 5,
           'n_prop_steps': 30,
            'n_ene_blocks': 1,
            'n_sr_blocks': 10,
            'n_blocks': 5,
            'n_walkers': 5,
            'seed': 98,
            'walker_type': 'rhf',
            'trial': 'rhf',
            'dt':0.005,
            'ad_mode':None,
            'use_gpu': False,
            }

u = lno_maker.thouless_trans(mycc.t1)
mo_t = mf.mo_coeff @ u
mf.mo_coeff = mo_t

from ad_afqmc import pyscf_interface, mpi_jax, driver
from mpi4py import MPI
#if not MPI.Is_finalized():
#    MPI.Finalize()

pyscf_interface.prep_afqmc(mf,chol_cut=1e-7)
ham_data, ham, prop, trial, wave_data, sampler, observable, options, _ \
    = (mpi_jax._prep_afqmc(options))

init_walkers = trial.get_init_walkers(wave_data, prop.n_walkers, restricted=True)
init_walkers = u.T @ init_walkers

driver.afqmc(ham_data,ham,prop,trial,wave_data,sampler,observable,options,MPI,init_walkers)
#INFO: ******************** input file end ********************


System: uname_result(system='Linux', node='yichi-thinkpad', release='4.4.0-26100-Microsoft', version='#5074-Microsoft Fri Jan 01 08:00:00 PST 2016', machine='x86_64')  Threads 12
Python 3.10.16 | packaged by conda-forge | (main, Dec  5 2024, 14:16:10) [GCC 13.3.0]
numpy 1.24.3  scipy 1.14.1  h5py 3.12.1
Date: Thu Sep 25 22:40:42 2025
PySCF version 2.8.0
PySCF path  /home/yichi/research/software/lno_pyscf
GIT HEAD (branch master) ef75f4190e4de208685670651dc6c467f72b6794

[ENV] PYSCF_EXT_PATH /home/yichi/research/software/pyscf
[CONFIG] conf_file None
[INPUT] verbose = 4
[INPUT] num. atoms = 10
[INPUT] num. electrons = 10
[INPUT] charge = 0
[INPUT] spin (= nelec alpha-beta = 2S) = 0
[INPUT] symmetry False subgroup None
[INPUT] Mole.unit = angstrom
[INPUT] Symbol           X                Y                Z      unit          X                Y                Z       unit  Magmom
[INPUT]  1 H      0.000000000000   0.000000000000   0.000000000000 AA    0.000000000000   0.000000000000   0.000000000000 Bohr   0.0
[INPUT]  2 H      1.000000000000   0.000000000000   0.000000000000 AA    1.889726124565   0.000000000000   0.000000000000 Bohr   0.0
[INPUT]  3 H      2.000000000000   0.000000000000   0.000000000000 AA    3.779452249130   0.000000000000   0.000000000000 Bohr   0.0
[INPUT]  4 H      3.000000000000   0.000000000000   0.000000000000 AA    5.669178373695   0.000000000000   0.000000000000 Bohr   0.0
[INPUT]  5 H      4.000000000000   0.000000000000   0.000000000000 AA    7.558904498260   0.000000000000   0.000000000000 Bohr   0.0
[INPUT]  6 H      5.000000000000   0.000000000000   0.000000000000 AA    9.448630622825   0.000000000000   0.000000000000 Bohr   0.0
[INPUT]  7 H      6.000000000000   0.000000000000   0.000000000000 AA   11.338356747390   0.000000000000   0.000000000000 Bohr   0.0
[INPUT]  8 H      7.000000000000   0.000000000000   0.000000000000 AA   13.228082871955   0.000000000000   0.000000000000 Bohr   0.0
[INPUT]  9 H      8.000000000000   0.000000000000   0.000000000000 AA   15.117808996520   0.000000000000   0.000000000000 Bohr   0.0
[INPUT] 10 H      9.000000000000   0.000000000000   0.000000000000 AA   17.007535121086   0.000000000000   0.000000000000 Bohr   0.0

nuclear repulsion = 10.2076604058814
number of shells = 10
number of NR pGTOs = 60
number of NR cGTOs = 10
basis = sto6g
ecp = {}
CPU time:         0.50


******** <class 'pyscf.scf.hf.RHF'> ********
method = RHF
initial guess = minao
damping factor = 0
level_shift factor = 0
DIIS = <class 'pyscf.scf.diis.CDIIS'>
diis_start_cycle = 1
diis_space = 8
diis_damp = 0
SCF conv_tol = 1e-09
SCF conv_tol_grad = None
SCF max_cycles = 50
direct_scf = True
direct_scf_tol = 1e-13
chkfile to save SCF result = /tmp/tmp8z10lwq9
max_memory 4000 MB (current use 82 MB)
Set gradient conv threshold to 3.16228e-05
Initial guess from minao.
init E= -4.17025168665218
  HOMO = -0.146073790841818  LUMO = 0.0736283907849704
cycle= 1 E= -5.23392110354241  delta_E= -1.06  |g|= 0.113  |ddm|= 2.26
  HOMO = -0.254530422268337  LUMO = 0.126647545276391
cycle= 2 E= -5.24517325422512  delta_E= -0.0113  |g|= 0.0462  |ddm|= 0.312
  HOMO = -0.261631890763272  LUMO = 0.14770559013236
cycle= 3 E= -5.24750174702562  delta_E= -0.00233  |g|= 0.0106  |ddm|= 0.157
  HOMO = -0.267607503512117  LUMO = 0.148479207487177
cycle= 4 E= -5.24761678043961  delta_E= -0.000115  |g|= 0.000907  |ddm|= 0.044
  HOMO = -0.26738553950481  LUMO = 0.148576614684717
cycle= 5 E= -5.24761731637648  delta_E= -5.36e-07  |g|= 0.000203  |ddm|= 0.0017
  HOMO = -0.26740923696354  LUMO = 0.148582853840369
cycle= 6 E= -5.2476173420352  delta_E= -2.57e-08  |g|= 2.18e-05  |ddm|= 0.000493
  HOMO = -0.267422025650931  LUMO = 0.148592685774515
cycle= 7 E= -5.24761734254922  delta_E= -5.14e-10  |g|= 1.67e-06  |ddm|= 9.44e-05
  HOMO = -0.267420755372815  LUMO = 0.148591935424545
Extra cycle  E= -5.24761734255151  delta_E= -2.3e-12  |g|= 6.73e-07  |ddm|= 3.75e-06
converged SCF energy = -5.24761734255151

******** <class 'pyscf.cc.ccsd.CCSD'> ********
CC2 = 0
CCSD nocc = 5, nmo = 10
max_cycle = 50
direct = 0
conv_tol = 1e-07
conv_tol_normt = 1e-05
diis_space = 6
diis_start_cycle = 0
diis_start_energy_diff = 1e+09
max_memory 4000 MB (current use 88 MB)
Init t2, MP2 energy = -5.35523747807868  E_corr(MP2) -0.107620135527168
Init E_corr(CCSD) = -0.107620135527455
cycle = 1  E_corr(CCSD) = -0.139531498119322  dE = -0.0319113626  norm(t1,t2) = 0.0748979
cycle = 2  E_corr(CCSD) = -0.153054640202571  dE = -0.0135231421  norm(t1,t2) = 0.0430034
cycle = 3  E_corr(CCSD) = -0.164615284156539  dE = -0.011560644  norm(t1,t2) = 0.0234584
cycle = 4  E_corr(CCSD) = -0.166182460909952  dE = -0.00156717675  norm(t1,t2) = 0.00737725
cycle = 5  E_corr(CCSD) = -0.16566519977735  dE = 0.000517261133  norm(t1,t2) = 0.00385594
cycle = 6  E_corr(CCSD) = -0.165759719402616  dE = -9.45196253e-05  norm(t1,t2) = 0.00105709
cycle = 7  E_corr(CCSD) = -0.165804173201118  dE = -4.44537985e-05  norm(t1,t2) = 0.000415662
cycle = 8  E_corr(CCSD) = -0.165776248545754  dE = 2.79246554e-05  norm(t1,t2) = 0.00016305
cycle = 9  E_corr(CCSD) = -0.165768881602527  dE = 7.36694323e-06  norm(t1,t2) = 5.05166e-05
cycle = 10  E_corr(CCSD) = -0.165770284594065  dE = -1.40299154e-06  norm(t1,t2) = 1.72054e-05
cycle = 11  E_corr(CCSD) = -0.165771348049684  dE = -1.06345562e-06  norm(t1,t2) = 7.52589e-06
cycle = 12  E_corr(CCSD) = -0.165771954938109  dE = -6.06888426e-07  norm(t1,t2) = 1.60611e-06
cycle = 13  E_corr(CCSD) = -0.165772040652599  dE = -8.57144896e-08  norm(t1,t2) = 5.36258e-07
CCSD converged
E(CCSD) = -5.413389383204113  E_corr = -0.1657720406525991
ccsd energy with T2=0 is:-5.2476384287008155
# Hostname: yichi-thinkpad
# System Type: Linux
# Machine Type: x86_64
# Processor: x86_64
#
# Preparing AFQMC calculation
# Calculating Cholesky integrals
# Finished calculating Cholesky integrals
#
# Size of the correlation space:
# Number of electrons: (5, 5)
# Number of basis functions: 10
# Number of Cholesky vectors: 27
#
# Number of MPI ranks: 1
#
# norb: 10
# nelec: (5, 5)
#
# n_eql: 5
# n_prop_steps: 30
# n_ene_blocks: 1
# n_sr_blocks: 10
# n_blocks: 5
# n_walkers: 5
# seed: 98
# walker_type: rhf
# trial: rhf
# dt: 0.005
# use_gpu: False
# n_exp_terms: 6
# orbital_rotation: True
# do_sr: True
# symmetry: False
# save_walkers: False
# ene0: 0.0
# free_projection: False
# n_batch: 1
# LNO: False
# orbE: 0
# maxError: 0.001
#
# Equilibration sweeps:
#   Iter        Block energy      Walltime
#     0      -5.247638366e+00     3.18e+00 
#     1      -5.414409637e+00     8.07e+00 
#     2      -5.386489868e+00     1.16e+01 
#     3      -5.413207531e+00     1.26e+01 
#     4      -5.375468731e+00     1.36e+01 
#     5      -5.412038803e+00     1.45e+01 
#
# Sampling sweeps:
#  Iter        Mean energy          Stochastic error       Walltime
     0      -5.395546436e+00                -              1.72e+01 
     1      -5.394613063e+00                -              1.74e+01 
     2      -5.393140421e+00                -              1.75e+01 
     3      -5.397998411e+00                -              1.76e+01 
     4      -5.390173316e+00        9.782324162e-03        1.77e+01 
#
# Number of large deviations: 0
# Number of outliers in post: 0 
#
# Mean: -5.39017332e+00
# Block size    # of blocks         Mean                Error
      1                5       -5.39017332e+00       9.782324e-03
      2                2       -5.39799841e+00       4.805445e-03
# Stocahstic error estimate: 9.782324e-03
#
AFQMC energy: -5.390 +/- 0.010

